### PR TITLE
feat: add overdetermined morphing

### PR DESCRIPTION
**Pull request by wangzhe04**
This pull request implements the following changes to morphing.py:

1.  Add the overdetermined morphing functionality which allows users to input more input as many basis points as possible.
2.  Add the additional input format that allows users to specify coupling for production, decay, and both separately
3.  Add an tutorial Jupyter Notebook for using morphing.py in the directory examples/morphing_example.ipynb
4.  Add additional tests for testing the implemented features in tests/test_morphing.py

### CODE REVIEW

**Comment by alexander-held** [comment]
Thanks for opening this! You can add the `sympy` dependency to `setup.cfg` here: https://github.com/madminer-tool/madminer/blob/dd223f8462120eb968b753499518dee6f229b052/setup.cfg#L28-L36. That should cause it to be picked up for the tests in continuous integration. The tests are probably going to fail without updates, so we'll also have to look into updating those and testing the new functionality.

There are currently a lot of empty lines getting removed, is this on purpose? It might be best to avoid this in this PR and do it separately in another one if it is a matter of just formatting.

Can you reproduce the issue you saw with `particle`? And do you mean `utils/particle.py` or some other file? If there is a conflict that can appear, it would be good to find a way to avoid this as well.

**Comment by wangzhe04** [comment]
Yes, I can reproduce the error, this is the error message i got after running a new runner file in the docker environement. 

Code: 
import morphing as m
if __name__=="__main__":
    a = m.PhysicsMorpher(parameter_max_power=[2,2])

Error:
root@976930b62f62:/madminer/madminer/madminer/utils# /bin/python3 /madminer/madminer/madminer/utils/non_runner.py
Traceback (most recent call last):
  File "/madminer/madminer/madminer/utils/non_runner.py", line 1, in <module>
    import morphing as m
  File "/madminer/madminer/madminer/utils/morphing.py", line 9, in <module>
    from madminer.models import Benchmark
  File "/usr/local/lib/python3.8/dist-packages/madminer/__init__.py", line 4, in <module>
    from .delphes import DelphesReader
  File "/usr/local/lib/python3.8/dist-packages/madminer/delphes/__init__.py", line 1, in <module>
    from .delphes_reader import DelphesReader
  File "/usr/local/lib/python3.8/dist-packages/madminer/delphes/delphes_reader.py", line 10, in <module>
    from madminer.utils.interfaces.delphes_root import parse_delphes_root_file
  File "/usr/local/lib/python3.8/dist-packages/madminer/utils/interfaces/delphes_root.py", line 11, in <module>
    from particle import Particle
  File "/madminer/madminer/madminer/utils/particle.py", line 4, in <module>
    from particle import Particle
ImportError: cannot import name 'Particle' from partially initialized module 'particle' (most likely due to a circular import) (/madminer/madminer/madminer/utils/particle.py)

I think it can be solved by changing the utils/particle.py to a new name such as utils/particle1.py

**Comment by alexander-held** [comment]
The issue you see is due to running from within the `madminer` library, you can resolve it by running your code from elsewhere, then
```python
from madminer.utils import morphing as m

if __name__ == "main":
    a = m.PhysicsMorpher(parameter_max_power=[2, 2])
```
works fine. In order to be able to import the code in this fashion, while still being able to change the `morphing` module and see the effects, you can do an [editable install](https://pip.pypa.io/en/stable/topics/local-project-installs/#editable-installs) e.g. via `pip install -e .` from within the `madminer` repository.

File: setup.cfg
Comment (matthewfeickert) [inline setup.cfg:37]: Is this version of `sympy` known to be the oldest version that will still work? Was part of the `sympy` API that is used in this PR added in `v1.10.1`?

Comment (wangzhe04) [reply setup.cfg:37]: It's the version that I used in my environment. I assume there are older versions that can work, but haven't tested yet. I'll try to test it sometime later.

Comment (matthewfeickert) [reply setup.cfg:37]: Yeah it would be good to not be overly restrictive when it comes to core dependencies.

File: madminer/utils/morphing.py
Comment (alexander-held) [inline madminer/utils/morphing.py:135]:
```suggestion
            Number of same couplings
```
We might need more explanations either here or elsewhere to clarify what this means.

Comment (wangzhe04) [reply madminer/utils/morphing.py]: Yes I'll add that it stands for both production and decay

**Comment by Sinclert** [comment]
👋🏻 Hi @wangzhe04

Thank you for your contributions, they are highly appreciated :)

Tuning in to ask for tests covering the modified [PhysicsMorpher](https://github.com/madminer-tool/madminer/blob/master/madminer/utils/morphing.py) functions before marking this PR ready to review (as of now: `find_components`, `set_basis` and `calculate_morphing_matrix`). The test suite is currently very limited, and increasing it will make development sustainable over time.

**Comment by wangzhe04** [comment]
> 👋🏻 Hi @wangzhe04
> 
> Thank you for your contributions, they are highly appreciated :)
> 
> Tuning in to ask for tests covering the modified [PhysicsMorpher](https://github.com/madminer-tool/madminer/blob/master/madminer/utils/morphing.py) functions before marking this PR ready to review (as of now: `find_components`, `set_basis` and `calculate_morphing_matrix`). The test suite is currently very limited, and increasing it will make development sustainable over time.

Thank you! Thank you for your contributions as well! 
I'll surely include some testings for these modified parts!

**Comment by matthewfeickert** [comment]
@alexander-held suggested in a meeting today that @wangzhe04 write some tests for the project that target the areas he's hoping to extend for his IRIS-HEP Fellow project in another PR. 
As @Sinclert has already pointed out, this will give the nice effect of both adding coverage in addition to giving a nice method to check to make sure that the changes @wangzhe04 makes are doing what are expected. :slightly_smiling_face: 

@wangzhe04 if you wanted to outline some of the ideas for this project in a GitHub Issue that might make it easier to step through some other parts later on in the review.

**Comment by wangzhe04** [comment]
> @alexander-held suggested in a meeting today that @wangzhe04 write some tests for the project that target the areas he's hoping to extend for his IRIS-HEP Fellow project in another PR. As @Sinclert has already pointed out, this will give the nice effect of both adding coverage in addition to giving a nice method to check to make sure that the changes @wangzhe04 makes are doing what are expected. 🙂
> 
> @wangzhe04 if you wanted to outline some of the ideas for this project in a GitHub Issue that might make it easier to step through some other parts later on in the review.

Sure, I'll look into creating a issue tomorrow, thanks!

**Comment by wangzhe04** [comment]
Changed to pull request#516

**Comment by Sinclert** [comment]
Hey @wangzhe04 .

Thanks again for all the effort you are putting into this. Having tests for such a big change is crucial :)

I am curious though, why are you moving changes to new branches, opening new PRs while closing previous ones? I am afraid conversation / context may be lost because of this _dance_.

**Comment by wangzhe04** [comment]
> Hey @wangzhe04 .
> 
> Thanks again for all the effort you are putting into this. Having tests for such a big change is crucial :)
> 
> I am curious though, why are you moving changes to new branches, opening new PRs while closing previous ones? I am afraid conversation / context may be lost because of this _dance_.

Oh, I saw the changes made in this PR are also showing in the new PR so I thought there is no need to keep this one. I'll reopen this one for reference. Thank you for pointing out.

**Comment by alexander-held** [comment]
I think this was a misunderstanding about the strategy, sorry I was probably not very clear about this. My proposal is getting tests for the existing functionality in first (which we can do via #516) and then use this PR to add the new functionality + expand the tests to cover the new additions.

File: madminer/utils/morphing.py
Comment (alexander-held) [inline madminer/utils/morphing.py:282]: `assert` will be disabled when Python is running with some optimizations, so if you want to use this as a way to inform the user, I would suggest raising an exception (e.g. `raise ValueError("some_text")`).

Comment (wangzhe04) [reply madminer/utils/morphing.py:282]: Sure, i will change asserts to exceptions

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:151]: Could you be more precise here? I am not sure what this _list_ is suppose to contain by its name.

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:423]: Even if we have not officially added Black as the project format checker, we should try to follow its style guide.

```suggestion
    def calculate_morphing_matrix(self, basis=None, gd=None, gp=None, gs=None):
```

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:160]: Please leave this empty line.

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:557]:
```suggestion
            q, r = np.linalg.qr(morphing_submatrix, 'reduced')
```

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:627]: This is occurrence number 1 of the _component weights calculation_ snippet.

Please extract these lines to a private method within this class.

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:654]: This is occurrence number 2 of the _component weights calculation_ snippet.

Please extract these lines to a private method within this class.

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:9]: I would rather have Sympy being imported as `sp`, instead of importing specific functions / modules directly. Reason being: `symbols` and `expand` as generic enough words that may collide with variable / function names  in the future.

Not sure if there is any reason why this should not be the case, I am not familiar with this library.

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:189]: If this loop is iterating until it finds _\<something\>_, shouldn't it have the `break` clause inside the if?

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:194]: If this loop is iterating until it finds _\<something\>_, shouldn't it have the `break` clause inside the if?

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:199]: If this loop is iterating until it finds _\<something\>_, shouldn't it have the `break` clause inside the if?

**Review by Sinclert — COMMENTED** [review]
I noticed the PR adding Morphing tests (https://github.com/madminer-tool/madminer/pull/516) was directly merged from its _draft_ state, so I decided to go ahead and provide feedback for this one despite its current state.

Once again, thanks for your contributions :)

**Comment by alexander-held** [comment]
> I noticed the PR adding Morphing tests (#516) was directly merged from its _draft_ state, so I decided to go ahead and provide feedback for this one despite its current state.

I unmarked from draft before looking over it again and decided to go ahead since it was a straightforward PR in my view. If you prefer, I am happy to always wait for your explicit approval instead @Sinclert. Either way, this PR is bigger and we were planning on sending a request once all changes are done (which should be within the next day or two I think).

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py:151]: Thank you for your feedbacks, i'll change it name to components in the next commit.

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py:423]: Thank you, I didn't run black in this commit but will run it before next commit.

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py:627]: Sure, I'll extract these to a private method

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py:199]: Yes, sorry for this mistake. I actually realized it when writing the test, I'll fix it in the next commit

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py:9]: yes i think it can be changed to import as sp

**Comment by wangzhe04** [comment]
This new commit should addresses issues mentioned above, please let me know if i missed anything or there are some other issues. 
Thank you all for your feedbacks.

File: examples/morphing_basis/morphing_example.ipynb
Comment (alexander-held) [inline examples/morphing_basis/morphing_example.ipynb:235]: I would suggest converting this to a comment field with a Markdown table since you do not use any of the content directly.

Comment (wangzhe04) [reply examples/morphing_basis/morphing_example.ipynb]: Sure, I changed it to the markdown table in the new commit

File: madminer/utils/morphing.py
Comment (alexander-held) [inline madminer/utils/morphing.py:185]: Is there a reason to reassign this variable name, or can we just keep using `BSM_max_power` here?

File: madminer/utils/morphing.py
Comment (alexander-held) [inline madminer/utils/morphing.py:137]:
```suggestion
            The number of parameters affecting the decay vertex.
```

File: madminer/utils/morphing.py
Comment (alexander-held) [inline madminer/utils/morphing.py:140]:
```suggestion
            The number of parameters affecting the production vertex.
```

File: madminer/utils/morphing.py
Comment (alexander-held) [inline madminer/utils/morphing.py:143]:
```suggestion
            The number of parameters that affect both the production and decay vertices.
```

File: madminer/utils/morphing.py
Comment (alexander-held) [inline madminer/utils/morphing.py:123]: Suggestion: I would find an order where `Nd` appears before `Np` more intuitive, as production happens before decay. It does not matter for the code, and users can specify the keyword arguments in the order they like, so I won't insist on this change.

File: madminer/utils/morphing.py
Comment (alexander-held) [inline madminer/utils/morphing.py:470]:
```suggestion
                Manually specified morphing basis of decay basis for which morphing matrix is calculated.
```

Comment (alexander-held) [reply madminer/utils/morphing.py:470]: (same for the other arguments below)

File: madminer/utils/morphing.py
Comment (alexander-held) [inline madminer/utils/morphing.py:622]: same comment as above about the order of production/decay

File: examples/morphing_basis/morphing_example.ipynb
Comment (alexander-held) [inline examples/morphing_basis/morphing_example.ipynb:50]: Please either uncomment this, or remove the lines if you think the information is not particularly interesting to put in the notebook.

File: examples/morphing_basis/morphing_example.ipynb
Comment (alexander-held) [inline examples/morphing_basis/morphing_example.ipynb:47]: Please try to keep the line lengths of these comments a bit shorter to avoid rendering issues.

File: examples/morphing_basis/morphing_example.ipynb
Comment (alexander-held) [inline examples/morphing_basis/morphing_example.ipynb:33]: You can turn this into a title by adding a `#` in front, I would suggest giving the whole notebook a title like `# Morphing examples` and then making this a subtitle like `## First Input Format`.

**Comment by alexander-held** [comment]
@wangzhe04 please also edit the very first message in this PR to include a brief high level summary of the changes here.

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py]: I changed it to keep using BSM_max_power in the new commit

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py]: Sure, I think I have changed all orders into Np, Nd, Ns in parameter inputs and comments in the new commit.

File: examples/morphing_basis/morphing_example.ipynb
Comment (wangzhe04) [reply examples/morphing_basis/morphing_example.ipynb]: I uncomment the printout in the new commit. I originally thought leaving this can let user to choose to check and compare the results

File: examples/morphing_basis/morphing_example.ipynb
Comment (wangzhe04) [reply examples/morphing_basis/morphing_example.ipynb]: Thank you, I added the general title and changed some texts to subtiltes

File: examples/morphing_basis/morphing_example.ipynb
Comment (wangzhe04) [reply examples/morphing_basis/morphing_example.ipynb]: thank you, I had reduced the lengths of many comments.

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:178]: Three comments here:

1. The iterator can be simplified: `range(0, len(...))` --> `range(len(...))` --> `enumerate(...)`
2. If you are iterating on an array just to access its elements (i.e. `mono[i]`), iterate through them directly
3. It seems the `components` array is not used, anywhere, before this point. Why not creating it here?

```suggestion
            components = [mono_value[0] for mono_value in mono]
```

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:187]: (Do not forget to remove variable `len_non_pmax` if this suggestion gets applied).

Similar to the feedback provided in a previous comment:

```suggestion
            for j, _ in enumerate(non_pmax_components):
```

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:1013]: No need to separate `np.array` conversion, and `return` clause in 2 different lines.

```suggestion
        return np.array(component_weights)
```

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py]: Thank you for the feedbacks! I've made the changes and also move the components down here.

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py]: Thank you, I have remove the variable len_non_pmax.

File: madminer/utils/morphing.py
Comment (Sinclert) [reply madminer/utils/morphing.py]: @wangzhe04 remember to delete [this line](https://github.com/wangzhe04/madminer/blob/d3caf7971155cda1cb7a23330e86ecce4ac4fd99/madminer/utils/morphing.py#L1009) above the one I suggested, as it is now useless.

Comment (wangzhe04) [reply madminer/utils/morphing.py]: Thanks for reminding me!, I have deleted in the new commit.

**Comment by wangzhe04** [comment]
In the new commit I included a function in the example notebook that it can check the minimum number of inputs requires. Should we include that in the morphing.py as well? It's a simple function so probably can be added the easily

**Comment by alexander-held** [comment]
That looks very useful to me! I think this would be a great addition to `morphing.py` in a new function (with a dedicated test for just this piece of functionality).

**Comment by wangzhe04** [comment]
I added get_min_basis function and the tests and also updated the morphing_example.ipynb to use the built in one. There was a mistake and some format issues in the test_morphing.py i first commited but it should be fixed now.

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:735]: This `if` - `else` block can be simplified. If you are adding the value `n_d` to `i` only when it is not zero, you might as well do it always, even if it is zero (_innocuous operation_) 🤷🏻‍♂️

```python
factor *= float(theta_p[i] ** self.components[c, i + n_d])
```

File: madminer/utils/morphing.py
Comment (Sinclert) [inline madminer/utils/morphing.py:746]: This whole `if` - `elif` -  `else` block can be simplified. If you are adding the values `n_d` and `n_p` to `k` only when they are not zero, you might as well do it always, even if they are zero (_innocuous operation_) 🤷🏻‍♂️

```python
factor *= float(theta_s[k] ** self.components[c, k + n_d + n_p])
```

File: tests/test_morphing.py
Comment (Sinclert) [inline tests/test_morphing.py:240]: Could you keep tuple style consistent?

```python
parameter_range=[
     (-1.0, 1.0),
     (-1.0, 1.0),
]
```

File: tests/test_morphing.py
Comment (Sinclert) [inline tests/test_morphing.py:260]: Keep tuple style consistent:

```python
parameter_range=[
     (-1.0, 1.0),
     (-1.0, 1.0),
]
```

**Review by Sinclert — COMMENTED** [review]
This is probably the last batch of requested changes! I will approve this PR as soon as they are addressed.

File: madminer/utils/morphing.py
Comment (wangzhe04) [reply madminer/utils/morphing.py]: Thank you for your feedback! I made this change and the below one in the new commit

File: tests/test_morphing.py
Comment (wangzhe04) [reply tests/test_morphing.py:254]: Thank you for your feedback, I changed these two input format in the new commit.

**Review by Sinclert — APPROVED** [review]
LGTM as it is! What do you think? @matthewfeickert @alexander-held

**Review by alexander-held — APPROVED** [review]
This looks good to me, I have no other comments. Thank you very much for getting this ready and all the interactions on the PR!

**Review by matthewfeickert — APPROVED** [review]
Taking a very quick look at this before lunch I think this is fine (added tests are very nice!). If I have any other comments I will open them up as separate Issues that can get resolved after this is merged. Thanks for the contributions @wangzhe04 and for @Sinclert and @alexander-held for the guidance here. :+1:

**Comment by alexander-held** [comment]
Thanks everyone! I'll leave it to @Sinclert to hit merge. For future reference, @wangzhe04 will present this work done as an IRIS-HEP fellow also at a meeting tomorrow: https://indico.cern.ch/event/1195271/.

**Pull request merged**
